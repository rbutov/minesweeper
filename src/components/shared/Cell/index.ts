enum CellTypes {
  CLOSED = 'closed',
  FLAG = 'flag',
  MINE = 'mine',
  MINE_RED = 'mine_red',
  MINE_WRONG = 'mine_wrong',
  TYPE_0 = 'type_0',
  TYPE_1 = 'type_1',
  TYPE_2 = 'type_2',
  TYPE_3 = 'type_3',
  TYPE_4 = 'type_4',
}

type CellType = `${CellTypes}`;

interface CellProps {
  type: CellType;
  row: number;
  col: number;
}

import closed from 'assets/cell/closed.svg';
import flag from 'assets/cell/flag.svg';
import mine from 'assets/cell/mine.svg';
import mine_red from 'assets/cell/mine_red.svg';
import mine_wrong from 'assets/cell/mine_wrong.svg';
import type0 from 'assets/cell/type0.svg';
import type1 from 'assets/cell/type1.svg';
import type2 from 'assets/cell/type2.svg';
import type3 from 'assets/cell/type3.svg';
import type4 from 'assets/cell/type4.svg';

const CellAsserts = {
  [CellTypes.CLOSED]: closed,
  [CellTypes.FLAG]: flag,
  [CellTypes.MINE]: mine,
  [CellTypes.MINE_RED]: mine_red,
  [CellTypes.MINE_WRONG]: mine_wrong,
  [CellTypes.TYPE_0]: type0,
  [CellTypes.TYPE_1]: type1,
  [CellTypes.TYPE_2]: type2,
  [CellTypes.TYPE_3]: type3,
  [CellTypes.TYPE_4]: type4,
};

export * from './Cell';
export { CellTypes, CellAsserts };
export type { CellProps, CellType };
