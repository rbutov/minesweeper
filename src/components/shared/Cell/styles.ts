import styled from 'styled-components';
import { CellAsserts, CellType, CellTypes } from '.';

const CellContainer = styled.div<{ type: CellType; failedMineKey: string | null }>`
  width: 24px;
  height: 24px;
  font-size: 10px;
  line-height: 23px;
  float: left;
  background-size: 100%;
  font-weight: 700;
  text-align: center;

  ${({ type }) => `background-image: url(${CellAsserts[type]})`};

  :active {
    ${({ type, failedMineKey }) =>
      type === CellTypes.CLOSED && !failedMineKey
        ? `background-image: url(${CellAsserts[CellTypes.TYPE_0]})`
        : ''};
  }
`;

export { CellContainer };
