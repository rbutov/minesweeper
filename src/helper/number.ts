const numberToArray = (num: number) => {
  const resArray = Array.from(String(num), Number);
  if (resArray.length < 3) {
    for (let i = 0; i < 4 - resArray.length; i++) {
      resArray.unshift(0);
    }
  }

  return resArray.map((n) => (isNaN(n) ? 10 : n));
};

export { numberToArray };
