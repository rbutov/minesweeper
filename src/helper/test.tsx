import React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';
import { ReactWrapper, ShallowWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { confStore } from 'store';

type Wrapper<P = unknown, S = unknown> = ShallowWrapper<P, S> | ReactWrapper<P, S>;
type MountFnType<P, S> = (
  node: React.ReactElement<P>,
  options?: Record<string, unknown>
) => Wrapper<P, S>;

const wrapAppComponent = <P, S>(
  renderer: MountFnType<P, S>,
  element: JSX.Element,
  other?: {
    store?: Store;
  }
): Wrapper<P, S> => {
  const coreStore = confStore();

  return renderer(<Provider store={other?.store ? other.store : coreStore}>{element}</Provider>);
};

const waitForComponentToPaint = async <P, S>(
  wrapper: Wrapper<P, S>,
  other?: { delay?: number; props?: Pick<P, keyof P> }
) => {
  await act(async () => {
    await runAllPromises();
    await new Promise((resolve) => setTimeout(resolve, other?.delay ? other.delay : 0));
    wrapper.setProps({ ...other?.props });
    wrapper.update();
  });
};

const runAllPromises = async () => {
  return new Promise<void>((resolve) => {
    setTimeout(resolve, 0);
  });
};

export { wrapAppComponent, waitForComponentToPaint };
export type { Wrapper };
