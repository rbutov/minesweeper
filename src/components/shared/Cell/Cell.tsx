import React from 'react';
import { useAppDispatch } from 'hooks/useAppDispatch';
import { revealKey, toggleFlag } from 'actions/fieldActions';
import { CellContainer } from './styles';
import { CellProps } from '.';

const Cell = ({ type, row, col, failedMineKey }: CellProps) => {
  const dispatch = useAppDispatch();

  const handleMouseDown = (ev: React.MouseEvent<HTMLElement>) => {
    ev.preventDefault();

    if (ev.button === 2) {
      dispatch(toggleFlag({ row, col }));
      return;
    }
  };

  const handleClick = (ev: React.MouseEvent<HTMLElement>) => {
    ev.preventDefault();
    dispatch(revealKey({ row, col }));
  };

  return (
    <CellContainer
      type={type}
      failedMineKey={failedMineKey}
      onMouseDown={handleMouseDown}
      onClick={handleClick}
    />
  );
};

export { Cell };
