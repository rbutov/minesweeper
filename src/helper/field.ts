import { Size } from 'types';
import { CellTypes } from 'components/shared/Cell';

interface ToKeyProps {
  row: number;
  col: number;
}

/**
 * _convert coordinates to key
 * @param row: _row number
 * @param col: _col number
 */
const toKey = ({ row, col }: ToKeyProps) => {
  return row + '-' + col;
};

const fromKey = (key: string) => {
  return key.split('-').map(Number);
};

interface GenerateMinesProps {
  rows: number;
  cols: number;
  minesCount: number;
}
const generateMines = ({ rows, cols, minesCount }: GenerateMinesProps) => {
  const mines = [];

  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      mines.push(toKey({ row, col }));
    }
  }
  mines.sort(() => (Math.random() > 0.5 ? 1 : -1));

  return mines.slice(0, minesCount);
};

interface GenerateMapProps {
  seedMines: string[];
  size: Size;
}
const generateMap = ({ seedMines, size }: GenerateMapProps) => {
  const map = new Map();

  const incrementDanger = (neighborKey: string) => {
    if (!map.has(neighborKey)) {
      map.set(neighborKey, 1);
    } else {
      const oldVal = map.get(neighborKey);

      if (oldVal !== 'mine') {
        map.set(neighborKey, oldVal + 1);
      }
    }
  };

  for (let key of seedMines) {
    map.set(key, 'mine');

    for (let neighborKey of getNeighbors({ key, size })) {
      incrementDanger(neighborKey);
    }
  }

  return Object.fromEntries(map);
};

interface isInBoundsProps {
  cell: number[];
  size: Size;
}
const isInBounds = ({ cell: [row, col], size: { rows, cols } }: isInBoundsProps) => {
  if (row < 0 || col < 0) {
    return false;
  }

  return !(row >= rows || col >= cols);
};

interface getNeighborsProps {
  key: string;
  size: Size;
}
const getNeighbors = ({ key, size }: getNeighborsProps) => {
  const [nRow, nCol] = fromKey(key);
  const neighborRowCols = [
    [nRow - 1, nCol - 1],
    [nRow - 1, nCol],
    [nRow - 1, nCol + 1],
    [nRow, nCol - 1],
    [nRow, nCol + 1],
    [nRow + 1, nCol - 1],
    [nRow + 1, nCol],
    [nRow + 1, nCol + 1],
  ];

  return neighborRowCols
    .filter((cell) => isInBounds({ cell, size }))
    .map(([row, col]) => toKey({ row, col }));
};

interface GetCellTypeProps {
  row: number;
  col: number;
  flaggedKeys: string[];
  revealedKeys: string[];
  map: Record<string, string | number> | null;
  failedMineKey: string | null;
}
const getCellType = ({
  row,
  col,
  flaggedKeys,
  revealedKeys,
  map,
  failedMineKey,
}: GetCellTypeProps) => {
  const key = toKey({ row, col });

  if (failedMineKey) {
    if (map && Object.prototype.hasOwnProperty.call(map, key) && map[key] === 'mine') {
      if (failedMineKey === key) {
        return CellTypes.MINE_RED;
      }
      if (flaggedKeys.includes(key)) {
        return CellTypes.FLAG;
      }

      return CellTypes.MINE;
    }

    if (
      flaggedKeys.includes(key) &&
      map &&
      Object.prototype.hasOwnProperty.call(map, key) &&
      map[key] !== 'mine'
    ) {
      return CellTypes.MINE_WRONG;
    }
  }

  if (revealedKeys.includes(key)) {
    if (map && Object.prototype.hasOwnProperty.call(map, key)) {
      switch (map[key]) {
        case 1:
          return CellTypes.TYPE_1;
        case 2:
          return CellTypes.TYPE_2;
        case 3:
          return CellTypes.TYPE_3;
        case 4:
          return CellTypes.TYPE_4;
        default:
          return CellTypes.MINE;
      }
    }

    return CellTypes.TYPE_0;
  }

  if (flaggedKeys.includes(key)) {
    return CellTypes.FLAG;
  }

  return CellTypes.CLOSED;
};

interface RevealCellProps {
  revealedKeys: Set<string>;
  map: Map<string, string | number>;
  key: string;
  size: Size;
}
const revealCell = ({ revealedKeys, map, key, size }: RevealCellProps) => {
  if (map.get(key) === 'mine') {
    return true;
  } else {
    propagateReveal({ revealedKeys, map, key, visited: new Set(), size });
  }

  return false;
};

interface PropagateRevealProps {
  revealedKeys: Set<string>;
  map: Map<string, string | number>;
  key: string;
  visited: Set<string>;
  size: Size;
}
const propagateReveal = ({ revealedKeys, map, key, visited, size }: PropagateRevealProps) => {
  revealedKeys.add(key);
  visited.add(key);

  const isEmpty = !map.has(key);
  if (isEmpty) {
    for (let neighborKey of getNeighbors({ key, size })) {
      if (!visited.has(neighborKey)) {
        propagateReveal({ revealedKeys, map, key: neighborKey, visited, size });
      }
    }
  }
};

interface GetAllMinesProps {
  map: Record<string, string | number> | null;
  minesCount: number;
}
const getAllMines = ({ map, minesCount }: GetAllMinesProps) => {
  const mines = [];

  for (let key in map) {
    if (map[key] === 'mine') {
      mines.push(key);
    }
    if (mines.length === minesCount) {
      break;
    }
  }

  return mines;
};

export { toKey, generateMines, generateMap, getCellType, revealCell, getAllMines };
