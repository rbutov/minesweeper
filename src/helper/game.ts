import { Faces } from 'components/shared/Face';
import { numberToArray } from 'helper/number';
import { Size } from 'types';

interface GetFaceProps {
  size: Size;
  minesCount: number;
  revealedKeys: string[];
  failedMineKey: string | null;
}

/**
 * _get face ENUM
 * @param failedMineKey: _explosed mine key
 * @param size: _field size
 * @param revealedKeys: _already revealed keys
 * @param minesCount: _mines count
 */
const getFace = ({ failedMineKey, size, revealedKeys, minesCount }: GetFaceProps) => {
  if (failedMineKey) {
    return Faces.LOSE;
  }
  if (size.cols * size.cols === revealedKeys.length - minesCount + 2) {
    return Faces.WIN;
  }

  return Faces.UNPRESSED;
};

interface GetMinesCountProps {
  minesCount: number;
  flaggedKeys: string[];
}

/**
 * _get rest of mines count (unflagged)
 * @param minesCount: _mines count
 * @param flaggedKeys: _flagged cells count
 */
const getMinesCount = ({ minesCount, flaggedKeys }: GetMinesCountProps) => {
  const diff = minesCount - flaggedKeys.length;

  return numberToArray(diff);
};

export { getFace, getMinesCount };
