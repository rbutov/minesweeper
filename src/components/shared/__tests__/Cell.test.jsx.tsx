import React from 'react';
import { wrapAppComponent } from 'helper/test';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Cell, CellTypes } from '../Cell';

describe('<Cell />', () => {
  const content = 'Test';

  it('render <Cell /> component', () => {
    const wrapper = wrapAppComponent(
      mount,
      <Cell row={0} col={0} failedMineKey={'0'} type={CellTypes.CLOSED} />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
