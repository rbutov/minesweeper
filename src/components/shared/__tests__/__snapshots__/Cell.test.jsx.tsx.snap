// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell /> render <Cell /> component 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Cell
    col={0}
    failedMineKey="0"
    row={0}
    type="closed"
  >
    <styled.div
      failedMineKey="0"
      onClick={[Function]}
      onMouseDown={[Function]}
      type="closed"
    >
      <div
        className="sc-bdvvtL htSWkN"
        onClick={[Function]}
        onMouseDown={[Function]}
        type="closed"
      />
    </styled.div>
  </Cell>
</Provider>
`;
